tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreStore.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (C) 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.runtimes.tomcat.session;
+
+import com.google.cloud.datastore.Blob;
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.DatastoreOptions;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.KeyQuery;
+import com.google.cloud.datastore.Query;
+import com.google.cloud.datastore.QueryResults;
+import com.google.common.collect.Streams;
+
+import org.apache.catalina.LifecycleException;
+import org.apache.catalina.Session;
+import org.apache.catalina.session.StandardSession;
+import org.apache.catalina.session.StoreBase;
+import org.apache.juli.logging.Log;
+import org.apache.juli.logging.LogFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.Arrays;
+
+/**
+ * This store interact with the Datastore service to persist sessions.
+ *
+ * <p>It does not make any assumptions about the manager, so it could be used
+ * by all manager implementations.</p>
+ *
+ * <p>However aggregations can be slow on the Datastore so if performance is a concern prefer using
+ * a manager implementation which is not using aggregations such as {@link DatastoreManager}</p>
+ */
+public class DatastoreStore extends StoreBase {
+
+  private static final Log log = LogFactory.getLog(DatastoreStore.class);
+
+  private Datastore datastore = null;
+  private KeyFactory keyFactory = null;
+  private KeyQuery.Builder keyQueryBuilder = null;
+
+  /**
+   * Name of the kind used in The Datastore for the session.
+   */
+  private String sessionKind = "GCloudSession";
```

Comment:
I would think no, in Tomcat we are serializing a org.apache.catalina.session.StandardSession whereas in Jetty a org.eclipse.jetty.server.session.SessionData is serialized.

It might be possible to have cross-server compatibility by relying on the standard HttpSession interface but that would involve change in both distributed sessions modules.